core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/policy/PolicyEngineImpl.java

```diff
@@ -0,0 +1,135 @@
+/*
+ *  Copyright (c) 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.contract.policy;
+
+import org.eclipse.dataspaceconnector.policy.engine.PolicyEvaluator;
+import org.eclipse.dataspaceconnector.policy.engine.RuleProblem;
+import org.eclipse.dataspaceconnector.policy.model.Duty;
+import org.eclipse.dataspaceconnector.policy.model.Permission;
+import org.eclipse.dataspaceconnector.policy.model.Policy;
+import org.eclipse.dataspaceconnector.policy.model.Prohibition;
+import org.eclipse.dataspaceconnector.policy.model.Rule;
+import org.eclipse.dataspaceconnector.spi.contract.agent.ParticipantAgent;
+import org.eclipse.dataspaceconnector.spi.contract.policy.AtomicConstraintFunction;
+import org.eclipse.dataspaceconnector.spi.contract.policy.PolicyContext;
+import org.eclipse.dataspaceconnector.spi.contract.policy.PolicyEngine;
+import org.eclipse.dataspaceconnector.spi.contract.policy.PolicyResult;
+import org.eclipse.dataspaceconnector.spi.contract.policy.RuleFunction;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Default implementation of the policy engine.
+ */
+public class PolicyEngineImpl implements PolicyEngine {
+    private List<ConstraintFunctionEntry<Rule>> constraintFunctions = new ArrayList<>();
+    private List<RuleFunctionEntry<Rule>> ruleFunctions = new ArrayList<>();
+    private List<BiFunction<Policy, PolicyContext, Boolean>> preValidators = new ArrayList<>();
+    private List<BiFunction<Policy, PolicyContext, Boolean>> postValidators = new ArrayList<>();
+
+    @Override
+    public PolicyResult evaluate(Policy policy, ParticipantAgent agent) {
+        var context = new PolicyContextImpl(agent);
+
+        for (BiFunction<Policy, PolicyContext, Boolean> validator : preValidators) {
+            if (!validator.apply(policy, context)) {
+                return new PolicyResult(context.hasProblems() ? context.getProblems() : List.of("Pre-validator failed: " + validator.getClass().getName()));
+            }
+        }
+
+        var evalBuilder = PolicyEvaluator.Builder.newInstance();
+        var evaluator = evalBuilder.build();
```

Comment:
suggest to move call to ```build()``` down to line 77. It works as it returns a reference but looks confusing.

Reply:
That was a cut-and-paste error...fixing it.